text-wrap balance should consider line-clamp when balancing

vitorroriz opened the pull request:
#### c1353df4a27c21ed9d969ee0f20a367a68ffbd29
<pre>
text-wrap balance should consider line-clamp when balancing
<a href="https://bugs.webkit.org/show_bug.cgi?id=268302">https://bugs.webkit.org/show_bug.cgi?id=268302</a>
<a href="https://rdar.apple.com/121858978">rdar://121858978</a>

Reviewed by Alan Baradlay.

According to spec resolution [1], if line-clamp
is defined, text-wrap: balance should balance
only within the clamped lines.

Up to this patch, we would balance taking into
consideration all the lines and we would clamp
it after balance.

This patches makes InlineContentBalancer::initialize()
take the maximum number of visible lines into account,
based into the line-clamp property.

Also, this allows for a small optimization:
If line-clamp clamps to 1 line, we can skip balacing.

[1] <a href="https://github.com/w3c/csswg-drafts/issues/9310">https://github.com/w3c/csswg-drafts/issues/9310</a>

* LayoutTests/TestExpectations:
* Source/WebCore/layout/formattingContexts/block/BlockLayoutState.h:
(WebCore::Layout::BlockLayoutState::LineClamp::allowedLineCount const):
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp:
(WebCore::Layout::InlineContentBalancer::initialize):
(WebCore::Layout::InlineContentBalancer::computeBalanceConstraints):
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.h:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::createDisplayContentForInlineContent):

Canonical link: <a href="https://commits.webkit.org/273800@main">https://commits.webkit.org/273800@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/867f21fe9f8ed303e857e435b8c1781dcf4088a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36708 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15647 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38941 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39354 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37952 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12756 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37271 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13183 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/11532 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/32721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40601 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33247 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/33072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/35592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8320 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12713 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
```diff
@@ -497,6 +497,13 @@ bool InlineFormattingContext::rebuildInlineItemListIfNeeded(const InlineDamage*
+    if (auto lineClamp = layoutState().parentBlockLayoutState().lineClamp())
+        return lineClamp->maximumLineCount > lineClamp->currentLineCount ? lineClamp->maximumLineCount - lineClamp->currentLineCount : 0;
+    return { };
+}
```
any particular reason why this has to be on InlineFormattingContext now?

vitorroriz replied:
It's just that this is already calculated at `InlineFormattingContext::createDisplayContentForInlineContent`, so I wanted to reuse it (Which I forgot to do in this patch).

alanbaradlay replied:
please don't. InlineFormattingContext's API should stay focused. if you are interested in sharing code, move it to a helper function to the util class but I don't think it's worth the effort.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp:
```diff
@@ -110,7 +110,8 @@ InlineContentBalancer::InlineContentBalancer(InlineFormattingContext& inlineForm
-    if (!m_inlineFormattingContext.layoutState().placedFloats().isEmpty()) {
+    auto numberOfVisibleLinesAllowed =  m_inlineFormattingContext.numberOfVisibleLinesAllowed();
+    if (!m_inlineFormattingContext.layoutState().placedFloats().isEmpty() || numberOfVisibleLinesAllowed == 1) {
         m_cannotBalanceContent = true;
```
I think this is quite misleading and somewhat error prone. numberOfVisibleLinesAllowed == 1 means we **_should not_** balance which is very different from cannot. -even if InlineContentBalancer currently ends up doing the same for both cases.

vitorroriz replied:
Agree. I've almost changed the name to `shouldNotBalanceContent`. I guess I'll do it. I assume that "should not" also makes sense for the case we just to avoid balancing for floats until it is unsupported.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.h:
```diff
@@ -62,6 +62,7 @@ class InlineFormattingContext {
     std::pair<LayoutUnit, LayoutUnit> minimumMaximumContentSize(const InlineDamage* = nullptr);
     LayoutUnit minimumContentSize(const InlineDamage* = nullptr);
     LayoutUnit maximumContentSize(const InlineDamage* = nullptr);
+    std::optional<size_t> numberOfVisibleLinesAllowed() const;
```
this should somehow reference to line-clamp to make it clear where this constrain comes from

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingUtils.cpp:
```diff
@@ -569,6 +569,13 @@ const InlineLayoutState& InlineFormattingUtils::layoutState() const
     return formattingContext().layoutState();
 }
 
+std::optional<size_t> InlineFormattingUtils::numberOfVisibleLinesAfterLineClamping() const
```
what does "after line clamping" mean here?

vitorroriz replied:
Hmmm, yeah, maybe it's not the best name. How about "numberOfVisibleLinesConsideringLineClamping"? I'm also open for suggestions of course!

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
```diff
@@ -345,17 +345,14 @@ void InlineFormattingContext::updateBoxGeometryForPlacedFloats(const LineLayoutR
-            return lineClamp->maximumLineCount > lineClamp->currentLineCount ? lineClamp->maximumLineCount - lineClamp->currentLineCount : 0;
-        return { };
-    }();
+    auto numberOfVisibleLinesConsideringLineClamp = formattingUtils().numberOfVisibleLinesConsideringLineClamp();
```
```
diff --git a/Source/WebCore/layout/formattingContexts/block/BlockLayoutState.h b/Source/WebCore/layout/formattingContexts/block/BlockLayoutState.h
index b3eb317d69f1..f59cdee0b941 100644
--- a/Source/WebCore/layout/formattingContexts/block/BlockLayoutState.h
+++ b/Source/WebCore/layout/formattingContexts/block/BlockLayoutState.h
@@ -37,6 +37,8 @@ class BlockFormattingContext;
 class BlockLayoutState {
 public:
     struct LineClamp {
+        size_t allowedLineCount() const { return std::max(maximumLineCount - currentLineCount, 0lu); }
+
         size_t maximumLineCount { 0 };
         size_t currentLineCount { 0 };
     };
diff --git a/Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp b/Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp
index 23a98ec5d724..df2a13e72dd6 100644
--- a/Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp
+++ b/Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp
@@ -345,11 +345,7 @@ void InlineFormattingContext::updateBoxGeometryForPlacedFloats(const LineLayoutR
 
 InlineRect InlineFormattingContext::createDisplayContentForInlineContent(const LineBox& lineBox, const LineLayoutResult& lineLayoutResult, const ConstraintsForInlineContent& constraints, InlineDisplay::Content& displayContent, size_t numberOfPreviousLinesWithInlineContent)
 {
-    auto numberOfVisibleLinesAllowed = [&] () -> std::optional<size_t> {
-        if (auto lineClamp = layoutState().parentBlockLayoutState().lineClamp())
-            return lineClamp->maximumLineCount > lineClamp->currentLineCount ? lineClamp->maximumLineCount - lineClamp->currentLineCount : 0;
-        return { };
-    }();
+    auto numberOfVisibleLinesAllowed = layoutState().parentBlockLayoutState().lineClamp() ? std::make_optional(layoutState().parentBlockLayoutState().lineClamp()->allowedLineCount()) : std::nullopt;

```
Maybe something like this would work and then you don't have to think about name at all (no util function is needed)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp:
```diff
@@ -110,8 +110,9 @@ InlineContentBalancer::InlineContentBalancer(InlineFormattingContext& inlineForm
+    auto numberOfVisibleLinesConsideringLineClamp = m_inlineFormattingContext.formattingUtils().numberOfVisibleLinesConsideringLineClamp();
+    if (!m_inlineFormattingContext.layoutState().placedFloats().isEmpty() || numberOfVisibleLinesConsideringLineClamp == 1) {
+        m_shouldNotBalanceContent = true;
         return;
```
What I really meant here was that there's a difference between cases when we bail out due to unable to balance (floats are present) and cases when there's only one line and we can take a fast path. InlineContentBalancer may not care about the difference yet but I think it would make the code read better if we decoupled these 2.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp:
```diff
@@ -157,14 +158,16 @@ void InlineContentBalancer::initialize()
         previousLine = PreviousLine { lineIndex, lineLayoutResult.contentGeometry.trailingOverflowingContentWidth, !lineLayoutResult.inlineContent.isEmpty() && lineLayoutResult.inlineContent.last().isLineBreak(), !lineLayoutResult.inlineContent.isEmpty(), lineLayoutResult.directionality.inlineBaseDirection, WTFMove(lineLayoutResult.floatContent.suspendedFloats) };
         lineIndex++;
+        if (numberOfVisibleLinesConsideringLineClamp && lineIndex >= numberOfVisibleLinesConsideringLineClamp)
+            break;
```
nit: you can bail out a lot earlier.

vitorroriz replied:
Don't we have to initialize the Balancer until we would reach the last line before clamping?

alanbaradlay replied:
when you say Balancer are you referring to InlineContentBalancer? -just wanna make sure we are talking about the same thing.

vitorroriz replied:
Yes, exactly!

alanbaradlay replied:
ok, yes we have to initialize the InlineContentBalancer, but we are _inside_ InlineContentBalancer. What I was trying to say with my original comment was please look and make sure we don't compute anything redundant here (i.e. perhaps moving this check a few lines above? -maybe "a lot earlier" was a bit of an exaggeration)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
```diff
@@ -345,17 +345,14 @@ void InlineFormattingContext::updateBoxGeometryForPlacedFloats(const LineLayoutR
-            return lineClamp->maximumLineCount > lineClamp->currentLineCount ? lineClamp->maximumLineCount - lineClamp->currentLineCount : 0;
-        return { };
-    }();
+    auto numberOfVisibleLinesConsideringLineClamp = layoutState().parentBlockLayoutState().lineClamp() ? std::make_optional(layoutState().parentBlockLayoutState().lineClamp()->allowedLineCount()) : std::nullopt;
```
do you mind keeping the current variable name? While it was not great as a function name (no context in the header) I think it's fine as a variable name here.

vitorroriz committed:
text-wrap balance should consider line-clamp when balancing
https://bugs.webkit.org/show_bug.cgi?id=268302
rdar://121858978

Reviewed by Alan Baradlay.

According to spec resolution [1], if line-clamp
is defined, text-wrap: balance should balance
only within the clamped lines.

Up to this patch, we would balance taking into
consideration all the lines and we would clamp
it after balance.

This patches makes InlineContentBalancer::initialize()
take the maximum number of visible lines into account,
based into the line-clamp property.

Also, this allows for a small optimization:
If line-clamp clamps to 1 line, we can skip balacing.

[1] https://github.com/w3c/csswg-drafts/issues/9310

* LayoutTests/TestExpectations:
* Source/WebCore/layout/formattingContexts/block/BlockLayoutState.h:
(WebCore::Layout::BlockLayoutState::LineClamp::allowedLineCount const):
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp:
(WebCore::Layout::InlineContentBalancer::initialize):
(WebCore::Layout::InlineContentBalancer::computeBalanceConstraints):
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.h:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::createDisplayContentForInlineContent):

Canonical link: https://commits.webkit.org/273800@main
